src/Currency.php

```diff
@@ -144,16 +145,31 @@ public function getAmountAsFractionalUnits(): int
     }
 
     /**
+     * @todo Use method getDecorationType() when decorating the result
      * {@inheritdoc}
      */
-    public function format(string $decimalsSeparator = '.', string $thousandsSeparator = '', int $extraPrecision = 0): string
+    public function format(CurrencyFormatInterface $currencyFormat = null): string
     {
-        $nDecimals = $this->currencyType->getNumFractionalDigits() + $extraPrecision;
+        if (is_null($currencyFormat)) {
+            $currencyFormat = new CurrencyFormat();
+        }
+
+        $precision = $currencyFormat->getPrecision();
+        if (is_null($precision)) {
+            $precision = $this->currencyType->getNumFractionalDigits();
+        }
+        $nDecimals = $precision + $currencyFormat->getExtraPrecision();
```

Comment:
This is one example of the composability that i referred in previous comments:

1. If the `$currencyFormat` instance returns a not-null value when we call `getPrecision`, the we should use this value.
2. If `getPrecision` returns null, then we should take the currency type precision.
3. If the `getExtraPrecision` method of the `$currencyFormat` instance returns a not-null value (AND the `$currencyFormat->getPrecision` call returned null), then, sum it to the currency type default precision.

The lines 153-161 don't match this expected behaviour.